imports/client/components/FancyEditor.tsx

```diff
@@ -0,0 +1,639 @@
+    // string, and whether that match was at the start of the field.
+    const displayName = u.displayName?.toLowerCase() ?? '';
+    const foundDisplayName = displayName.includes(needle);
+    const startsDisplayName = displayName.startsWith(needle);
+    const emails = u.emails?.map((addrObj) => addrObj.address.toLowerCase()) ?? [];
+    const foundEmail = emails.some((addr) => addr.includes(needle));
+    const startsEmail = emails.some((addr) => addr.startsWith(needle));
+    const googEmail = u.googleAccount?.toLowerCase() ?? '';
+    const foundGoogle = googEmail.includes(needle);
+    const startsGoogle = googEmail.startsWith(needle);
+    const discordName = u.discordAccount ? `${u.discordAccount.username.toLowerCase()}#${u.discordAccount.discriminator}` : '';
+    const foundDiscord = discordName.includes(needle);
+    const startsDiscord = discordName.startsWith(needle);
+
+    return {
+      ...u,
+      foundDisplayName,
+      startsDisplayName,
+      foundEmail,
+      startsEmail,
+      foundGoogle,
+      startsGoogle,
+      foundDiscord,
+      startsDiscord,
+    };
+  });
+
+  // Only display those users which matched at least one field.
+  const augmentedMatches = augmentedUsers.filter((u) => {
+    return u.foundDisplayName || u.foundEmail || u.foundGoogle || u.foundDiscord;
+  });
+
+  // Sort the remaining users by relevancy, with extra weight given to
+  // matching at the start of the field, and with display name more important
+  // that google email address or discord username#discriminator.
+  // Note that `users` was originally sorted by displayName and that
+  // `sortedBy` is a stable sort.
+  return sortedBy(augmentedMatches, (u) => {
+    return (
+      (u.startsDisplayName ? -10000 : 0) +
+      (u.startsGoogle || u.startsEmail || u.startsDiscord ? -5000 : 0) +
+      (u.foundDisplayName ? -100 : 0) +
+      (u.foundEmail ? -50 : 0) +
+      (u.foundGoogle ? -10 : 0) +
+      (u.foundDiscord ? -5 : 0)
+    );
+  });
+}
+
+const StyledMessage = styled.p`
+  margin-bottom: 0;
+`;
+
+const AutocompleteContainer = styled.div`
+  position: absolute;
+  z-index: 6;
+  padding: 3px;
+  background: white;
+  border-radius: 4px;
+  box-shadow: 0 1px 5px rgb(0 0 0 / 20%);
+`;
+
+type TraverseCallback = (token: marked.Token, offset: number) => void;
+
+// Walks the tokens provided, calling callback, and keeping track of the raw offset
+// into the input string along the way.
+const walkTokenList = (
+  tokens: marked.Token[],
+  callback: TraverseCallback,
+  offset: number,
+) => {
+  const start = offset;
+  let end = start;
+  tokens.forEach((token) => {
+    // mutual recursion requires this
+    // eslint-disable-next-line @typescript-eslint/no-use-before-define
+    walkToken(token, callback, end);
+    end += token.raw.length;
+  });
+};
+
+const walkToken = (token: marked.Token, callback: TraverseCallback, offset: number) => {
+  if ((token as any).tokens) {
```

Comment:
FTR, I think you can phrase this as `'tokens' in token && token.tokens` to avoid the `any` cast

Reply:
Ah, neat. Fixed in 7cbd867e0b78b131bd0536cd99367ba791475764